### ISSUE TITLE
Should be able to infer type for addEventListener argument

### ISSUE THREAD

**Issue by zapashcanon** [facebook/flow#7380]
Hi,

In the following:

```js
const f = function (e) {
  switch (e.key) {
    case "e":
      return 1;
  };
};

document.addEventListener("keydown", f);
```

I get the following:

```txt
Could not decide which case to select. Since case 3 [1] may work but if it doesn't case 9 [2] looks promising too. To
fix add a type annotation to e [3].

 [3]  78│ const f = function (e) {
        :
      94│   };
      95│ };
      96│
      97│ document.addEventListener("keydown", f);

 [1] 157│   addEventListener(type: KeyboardEventTypes, listener: KeyboardEventListener, optionsOrUseCapture?: EventListenerOptionsOrUseCapture): void;
        :
 [2] 163│   addEventListener(type: string, listener: EventListener, optionsOrUseCapture?: EventListenerOptionsOrUseCapture): void;

Found 1 error
```

Flow should be able to infer that I'm using `KeybiardEventTypes` and not `string`, as I'm using `e.key` which is only defined for the first one.

**Comment by zapashcanon** [comment]
In the case of `KeyboardEventTypes`, then, `e` is going to be a `KeyboardEvent`, and thus will have a `key` prop.

I'm not sure it will have a `key` prop. in the case of `string`, but:

- if not, then Flow should be able to infer that I'm using `KeyboardEvenType`
- if yes, then, there's no issue as it'll always work, and Flow should consider the function to be polymorphic

**Comment by jbrown215** [comment]
Ah, I must have misunderstood before. 
In general, union speculation and determining which branch to go down is difficult. When Flow is unsure, it asks for an annotation. I'm not familiar enough with union speculation to tell you if this is a bug or not. Does adding an annotation to `e` fix your issue?

**Comment by zapashcanon** [comment]
> Does adding an annotation to `e` fix your issue?

Yes, I fixed it this way even before opening the issue, but I think it could be possible to avoid having to add it.
